File: modules/TechSpecGenerator/src/api_doc_processor/nodes/llm_processing_node.py
"""LLM processing node for the workflow."""

from rich.console import Console
from rich.progress import Progress, SpinnerColumn, TextColumn

from ..workflow_state import WorkflowState
from ..llm_client import LLMClient

console = Console()


def llm_processing_node(state: WorkflowState) -> WorkflowState:
    """
    Generate technical specification using LLM.
    
    Args:
        state: Current workflow state
        
    Returns:
        Updated state with generated tech spec
    """
    if not state["markdown_files"]:
        state["errors"].append("No markdown files to process")
        console.print("[red]Error:[/red] No markdown files to process")
        return state
    
    console.print(f"\n[bold]Step 2: Generating technical specification...[/bold]")
    
    # Initialize LLM client
    try:
        llm_client = LLMClient(state["config"].litellm)
    except Exception as e:
        error_msg = f"Failed to initialize LLM client: {str(e)}"
        state["errors"].append(error_msg)
        console.print(f"[red]Error:[/red] {error_msg}")
        return state
    
    # Show token estimation
    try:
        token_estimate = llm_client.estimate_token_usage(state["markdown_files"])
        if "error" not in token_estimate:
            state["metadata"]["estimated_tokens"] = token_estimate
            console.print(f"[dim]Estimated tokens: ~{token_estimate['estimated_input_tokens']} input + {token_estimate['max_output_tokens']} output[/dim]")
    except Exception as e:
        state["warnings"].append(f"Token estimation failed: {str(e)}")
    
    # Generate tech spec
    try:
        with Progress(
            SpinnerColumn(),
            TextColumn("[progress.description]{task.description}"),
            console=console,
        ) as progress:
            llm_task = progress.add_task("Generating tech spec...", total=None)
            
            spec_success, tech_spec, spec_error = llm_client.generate_tech_spec(
                state["markdown_files"], state["config"].prompt
            )
            
            if spec_success:
                # Save the tech spec
                specs_dir = state["output_dir"] / "specs"
                specs_dir.mkdir(exist_ok=True)
                spec_filepath = llm_client.save_tech_spec(tech_spec, specs_dir)
                
                # Update state
                state["tech_spec"] = tech_spec
                state["spec_filepath"] = spec_filepath
                state["metadata"]["spec_generated"] = True
                
                progress.update(llm_task, description="Tech spec generated!")
                console.print(f"\n[green]✓[/green] Technical specification generated!")
                console.print(f"[dim]Saved to: {spec_filepath}[/dim]")
            else:
                state["errors"].append(f"Tech spec generation failed: {spec_error}")
                state["metadata"]["spec_generated"] = False
                progress.update(llm_task, description="Tech spec generation failed!")
                console.print(f"\n[red]Error generating tech spec:[/red] {spec_error}")
    
    except Exception as e:
        error_msg = f"Error during tech spec generation: {str(e)}"
        state["errors"].append(error_msg)
        state["metadata"]["spec_generated"] = False
        console.print(f"\n[red]Error:[/red] {error_msg}")
    
    return state

File: modules/TechSpecGenerator/src/nodes/llm-processing-node.ts
/**
 * LLM Processing Node - TypeScript implementation
 */

import * as fs from 'fs-extra';
import * as path from 'path';
import { OpenAI } from 'openai';
import { WorkflowState } from '../types/workflow-state';
import { console } from '../utils/console';
import { Progress } from '../utils/progress';

export async function llmProcessingNode(state: WorkflowState): Promise<WorkflowState> {
  console.print('\n[bold]Step 3: LLM Processing...[/bold]');
  
  if (state.markdown_files.length === 0) {
    console.warn('No markdown files to process');
    return state;
  }
  
  const progress = new Progress();
  const newState = { ...state };
  
  try {
    // Read all markdown files
    const taskId = progress.addTask('Reading markdown files');
    const markdownContent = await readMarkdownFiles(state.markdown_files);
    progress.completeTask(taskId, `Read ${state.markdown_files.length} markdown files`);
    
    // Generate technical specification
    const specTaskId = progress.addTask('Generating technical specification with AI');
    const techSpec = await generateTechSpec(markdownContent, state.config);
    progress.completeTask(specTaskId, 'Technical specification generated');
    
    // Save tech spec
    const saveTaskId = progress.addTask('Saving technical specification');
    const specsDir = path.join(state.output_dir, 'specs');
    await fs.ensureDir(specsDir);
    
    const specFilepath = path.join(specsDir, 'technical_specification.md');
    await fs.writeFile(specFilepath, techSpec, 'utf-8');
    progress.completeTask(saveTaskId, `Specification saved to ${specFilepath}`);
    
    // Update state
    newState.tech_spec = techSpec;
    newState.spec_filepath = specFilepath;
    newState.metadata.spec_generated = true;
    
    // Estimate tokens (simplified calculation)
    newState.metadata.estimated_tokens = {
      input: Math.ceil(markdownContent.length / 4), // Rough estimate
      output: Math.ceil(techSpec.length / 4)
    };
    
    console.success('LLM processing completed successfully');
    
  } catch (error) {
    const errorMsg = `LLM processing failed: ${error}`;
    console.error(errorMsg);
    newState.errors.push(errorMsg);
    newState.metadata.spec_generated = false;
  }
  
  return newState;
}

async function readMarkdownFiles(filePaths: string[]): Promise<string> {
  const contents: string[] = [];
  
  for (const filePath of filePaths) {
    try {
      const content = await fs.readFile(filePath, 'utf-8');
      contents.push(`\n\n## Content from ${path.basename(filePath)}\n\n${content}`);
    } catch (error) {
      console.warn(`Failed to read ${filePath}: ${error}`);
    }
  }
  
  return contents.join('\n');
}

async function generateTechSpec(content: string, config: any): Promise<string> {
  const prompt = `
Analyze the following API documentation content and create a comprehensive technical specification.

Focus on:
1. API endpoints and their methods
2. Request/response schemas
3. Authentication mechanisms
4. Error handling
5. Rate limiting
6. Data models

Content:
${content}

Generate a detailed technical specification in markdown format that can be used to create a mock server.
  `.trim();

  try {
    // Initialize OpenAI (assuming OpenAI-compatible API)
    const openai = new OpenAI({
      apiKey: config.litellm.api_key,
      baseURL: config.litellm.base_url
    });

    const response = await openai.chat.completions.create({
      model: config.litellm.model,
      messages: [
        {
          role: 'user',
          content: prompt
        }
      ],
      max_tokens: config.litellm.max_tokens,
      temperature: 0.7
    });

    const techSpec = response.choices[0]?.message?.content;
    
    if (!techSpec) {
      throw new Error('No response from LLM');
    }
    
    return techSpec;
    
  } catch (error: any) {
    throw new Error(`LLM API request failed: ${error.message}`);
  }
}